File: zenoh_pc_nodes/python/continuous_stress_validator.py
#!/usr/bin/env python3
"""
continuous_stress_validator.py — 2-Minute High-Frequency Continuous Stress Validator (Python)
==========================================================================================
Validates continuous streaming on 34 topics + frequent service calls (every 2s) for 120s.
"""

import sys, os, time, threading
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../../shared_libraries/python')))

from zenoh_ros import ZenohNode, ZenohConfig
from zenoh_ros.std_msgs import (
    z_Bool, z_Int8, z_UInt8, z_Int16, z_UInt16, z_Int32, z_UInt32, z_Int64, z_UInt64,
    z_Float32, z_Float64, z_String, z_Header, z_Int32MultiArray, z_Float64MultiArray
)
from zenoh_ros.sensor_msgs import z_Imu, z_Temperature, z_Range, z_RelativeHumidity, z_BatteryState, z_NavSatFix, z_JoyFeedback, z_JointState
from zenoh_ros.geometry_msgs import z_Vector3, z_Quaternion, z_Pose, z_Twist

from zenoh_ros.custom_msgs import (
    z_SetLED, z_MPU6050Data, z_RobotState, z_MotorStatus, z_SensorTelemetry, z_RobotDiagnostic, z_ComplexRobotState
)
from zenoh_ros.custom_srvs import z_SetLEDColor, z_SetColor, z_ConfigureRobot, z_FullSystemControl

GRN = "\033[32m"; RED = "\033[31m"; YLW = "\033[33m"; CYN = "\033[36m"; BOLD = "\033[1m"; RST = "\033[0m"

topic_counts = {}
lock = threading.Lock()

def make_cb(topic_name):
    def cb(msg):
        with lock:
            topic_counts[topic_name] = topic_counts.get(topic_name, 0) + 1
    return cb

class ContinuousStressNode(ZenohNode):
    def __init__(self):
        super().__init__("continuous_stress_node")
        print(f"\n{CYN}{BOLD}╔══════════════════════════════════════════════════════════════╗")
        print(f"║  CONTINUOUS 2-MINUTE STRESS TEST & VALIDATOR (Python)       ║")
        print(f"║  Testing continuous data stream & frequent service calls     ║")
        print(f"╚══════════════════════════════════════════════════════════════╝{RST}\n")

        std_topics = [
            (z_Bool, "test/bool"), (z_Int8, "test/int8"), (z_UInt8, "test/uint8"),
            (z_Int16, "test/int16"), (z_UInt16, "test/uint16"), (z_Int32, "test/int32"),
            (z_UInt32, "test/uint32"), (z_Int64, "test/int64"), (z_UInt64, "test/uint64"),
            (z_Float32, "test/float32"), (z_Float64, "test/float64"), (z_String, "test/string"),
            (z_Header, "test/header"), (z_Int32MultiArray, "test/int32_array"),
            (z_Float64MultiArray, "test/float64_array"), (z_Imu, "test/imu"),
            (z_Temperature, "test/temperature"), (z_Range, "test/range"),
            (z_RelativeHumidity, "test/humidity"), (z_BatteryState, "test/battery"),
            (z_NavSatFix, "test/navsatfix"), (z_JoyFeedback, "test/joyfeedback"),
            (z_JointState, "test/jointstate"), (z_Vector3, "test/vector3"),
            (z_Quaternion, "test/quaternion"), (z_Pose, "test/pose"), (z_Twist, "test/twist")
        ]

        custom_topics = [
            (z_SetLED, "test_custom/set_led"), (z_MPU6050Data, "test_custom/mpu6050"),
            (z_RobotState, "test_custom/robot_state"), (z_MotorStatus, "test_custom/motor_status"),
            (z_SensorTelemetry, "test_custom/telemetry"), (z_RobotDiagnostic, "test_custom/robot_diag"),
            (z_ComplexRobotState, "test_custom/complex_state")
        ]

        for cls, top in std_topics + custom_topics:
            self.z_create_subscription(cls, top, make_cb(top))

        self.cli_set_led_clr = self.z_create_client(z_SetLEDColor, "srv_custom/set_led_color")
        self.cli_set_clr     = self.z_create_client(z_SetColor, "srv_custom/set_color")
        self.cli_cfg_robot   = self.z_create_client(z_ConfigureRobot, "srv_custom/configure_robot")
        self.cli_full_ctrl   = self.z_create_client(z_FullSystemControl, "srv_custom/full_system_control")

def main():
    config = ZenohConfig(host="10.42.0.50", port=7447)
    ZenohNode.init(config)
    node = ContinuousStressNode()

    t = threading.Thread(target=node.z_spin, daemon=True)
    t.start()

    start_time = time.time()
    duration = 120 # 2 minutes
    last_srv_time = time.time()

    total_srv_calls = 0
    successful_srv_calls = 0

    print(f"{YLW}Starting 120-second continuous stress test...{RST}")

    while time.time() - start_time < duration:
        time.sleep(0.5)
        elapsed = int(time.time() - start_time)

        # Service call round every 2 seconds
        if time.time() - last_srv_time >= 2.0:
            last_srv_time = time.time()
            total_srv_calls += 4

            try:
                res1 = node.cli_set_led_clr.call(z_SetLEDColor.Request(led_data=z_SetLED(r=255, g=0, b=0, brightness=100, led_num=1)), timeout_sec=2.0)
                if res1 and res1.success: successful_srv_calls += 1
                time.sleep(0.05)

                res2 = node.cli_set_clr.call(z_SetColor.Request(r=100, g=200, b=50), timeout_sec=2.0)
                if res2 and res2.success: successful_srv_calls += 1
                time.sleep(0.05)

                req3 = z_ConfigureRobot.Request(target_status=z_MotorStatus(motor_id=1, speed=120.0, is_active=True), mode=1)
                res3 = node.cli_cfg_robot.call(req3, timeout_sec=2.0)
                if res3 and res3.success: successful_srv_calls += 1
                time.sleep(0.05)

                req4 = z_FullSystemControl.Request(target_state=z_ComplexRobotState(robot_mode="STRESS", cycle_count=elapsed), command_code=7)
                res4 = node.cli_full_ctrl.call(req4, timeout_sec=2.0)
                if res4 and res4.ack: successful_srv_calls += 1
            except Exception as e:
                print(f"  {RED}⚠ Service call exception: {e}{RST}")

        with lock:
            rcvd_topics = len(topic_counts)
            total_msgs = sum(topic_counts.values())

        if elapsed % 10 == 0:
            print(f"  [{elapsed:3d}s/120s] Topics Received: {rcvd_topics:2d}/34 | Total Msgs: {total_msgs:5d} | Srv Calls OK: {successful_srv_calls}/{total_srv_calls}")

    print(f"\n{BOLD}{'═'*64}")
    print(f"  CONTINUOUS STRESS TEST FINISHED ({int(time.time() - start_time)}s elapsed)")
    print(f"  Topics Active  : {len(topic_counts)}/34")
    print(f"  Total Messages : {sum(topic_counts.values())}")
    print(f"  Service Calls  : {successful_srv_calls}/{total_srv_calls} PASSED ({(successful_srv_calls/total_srv_calls)*100:.1f}%)")
    if len(topic_counts) == 34 and successful_srv_calls == total_srv_calls:
        print(f"  {GRN}{BOLD}✓ 100% PERFECT CONTINUOUS STABILITY VERIFIED! ZERO LOSS!{RST}")
    else:
        print(f"  {RED}{BOLD}⚠  STRESS TEST COMPLETED WITH LOSS/ERRORS{RST}")
    print(f"{'═'*64}{RST}\n")

    node.z_destroy()

if __name__ == "__main__":
    main()

File: zenoh_pc_nodes/cpp/src/continuous_stress_validator.cpp
/**
 * continuous_stress_validator.cpp — 2-Minute High-Frequency Continuous Stress Validator (C++)
 * =========================================================================================
 * Validates continuous streaming on 34 topics + frequent service calls (every 2s) for 120s.
 */

#include <iostream>
#include <chrono>
#include <thread>
#include <map>
#include <mutex>
#include <atomic>

#include <zenoh_ros/ZenohRosPC.h>

// Standard ROS 2 Includes
#include <zenoh_ros/std_msgs/z_Bool.h>
#include <zenoh_ros/std_msgs/z_Int8.h>
#include <zenoh_ros/std_msgs/z_UInt8.h>
#include <zenoh_ros/std_msgs/z_Int16.h>
#include <zenoh_ros/std_msgs/z_UInt16.h>
#include <zenoh_ros/std_msgs/z_Int32.h>
#include <zenoh_ros/std_msgs/z_UInt32.h>
#include <zenoh_ros/std_msgs/z_Int64.h>
#include <zenoh_ros/std_msgs/z_UInt64.h>
#include <zenoh_ros/std_msgs/z_Float32.h>
#include <zenoh_ros/std_msgs/z_Float64.h>
#include <zenoh_ros/std_msgs/z_String.h>
#include <zenoh_ros/std_msgs/z_Header.h>
#include <zenoh_ros/std_msgs/z_Int32MultiArray.h>
#include <zenoh_ros/std_msgs/z_Float64MultiArray.h>
#include <zenoh_ros/sensor_msgs/z_Imu.h>
#include <zenoh_ros/sensor_msgs/z_Temperature.h>
#include <zenoh_ros/sensor_msgs/z_Range.h>
#include <zenoh_ros/sensor_msgs/z_RelativeHumidity.h>
#include <zenoh_ros/sensor_msgs/z_BatteryState.h>
#include <zenoh_ros/sensor_msgs/z_NavSatFix.h>
#include <zenoh_ros/sensor_msgs/z_JoyFeedback.h>
#include <zenoh_ros/sensor_msgs/z_JointState.h>
#include <zenoh_ros/geometry_msgs/z_Vector3.h>
#include <zenoh_ros/geometry_msgs/z_Quaternion.h>
#include <zenoh_ros/geometry_msgs/z_Pose.h>
#include <zenoh_ros/geometry_msgs/z_Twist.h>

// Custom Interface Includes
#include <zenoh_ros/custom_msgs/z_SetLED.h>
#include <zenoh_ros/custom_msgs/z_MPU6050Data.h>
#include <zenoh_ros/custom_msgs/z_RobotState.h>
#include <zenoh_ros/custom_msgs/z_MotorStatus.h>
#include <zenoh_ros/custom_msgs/z_SensorTelemetry.h>
#include <zenoh_ros/custom_msgs/z_RobotDiagnostic.h>
#include <zenoh_ros/custom_msgs/z_ComplexRobotState.h>

#include <zenoh_ros/custom_srvs/z_SetLEDColor.h>
#include <zenoh_ros/custom_srvs/z_SetColor.h>
#include <zenoh_ros/custom_srvs/z_ConfigureRobot.h>
#include <zenoh_ros/custom_srvs/z_FullSystemControl.h>

std::map<std::string, uint64_t> topic_counts;
std::mutex mtx;
std::atomic<uint64_t> total_msg_counter{0};

void record_msg(const std::string& top) {
    std::lock_guard<std::mutex> lock(mtx);
    topic_counts[top]++;
    total_msg_counter++;
}

class ContinuousStressNode : public ZenohNode {
public:
    ContinuousStressNode() : ZenohNode("continuous_stress_cpp_node") {
        z_create_subscription<z_Bool>("test/bool", [](const z_Bool&) { record_msg("test/bool"); });
        z_create_subscription<z_Int8>("test/int8", [](const z_Int8&) { record_msg("test/int8"); });
        z_create_subscription<z_UInt8>("test/uint8", [](const z_UInt8&) { record_msg("test/uint8"); });
        z_create_subscription<z_Int16>("test/int16", [](const z_Int16&) { record_msg("test/int16"); });
        z_create_subscription<z_UInt16>("test/uint16", [](const z_UInt16&) { record_msg("test/uint16"); });
        z_create_subscription<z_Int32>("test/int32", [](const z_Int32&) { record_msg("test/int32"); });
        z_create_subscription<z_UInt32>("test/uint32", [](const z_UInt32&) { record_msg("test/uint32"); });
        z_create_subscription<z_Int64>("test/int64", [](const z_Int64&) { record_msg("test/int64"); });
        z_create_subscription<z_UInt64>("test/uint64", [](const z_UInt64&) { record_msg("test/uint64"); });
        z_create_subscription<z_Float32>("test/float32", [](const z_Float32&) { record_msg("test/float32"); });
        z_create_subscription<z_Float64>("test/float64", [](const z_Float64&) { record_msg("test/float64"); });
        z_create_subscription<z_String>("test/string", [](const z_String&) { record_msg("test/string"); });
        z_create_subscription<z_Header>("test/header", [](const z_Header&) { record_msg("test/header"); });
        z_create_subscription<z_Int32MultiArray>("test/int32_array", [](const z_Int32MultiArray&) { record_msg("test/int32_array"); });
        z_create_subscription<z_Float64MultiArray>("test/float64_array", [](const z_Float64MultiArray&) { record_msg("test/float64_array"); });
        z_create_subscription<z_Imu>("test/imu", [](const z_Imu&) { record_msg("test/imu"); });
        z_create_subscription<z_Temperature>("test/temperature", [](const z_Temperature&) { record_msg("test/temperature"); });
        z_create_subscription<z_Range>("test/range", [](const z_Range&) { record_msg("test/range"); });
        z_create_subscription<z_RelativeHumidity>("test/humidity", [](const z_RelativeHumidity&) { record_msg("test/humidity"); });
        z_create_subscription<z_BatteryState>("test/battery", [](const z_BatteryState&) { record_msg("test/battery"); });
        z_create_subscription<z_NavSatFix>("test/navsatfix", [](const z_NavSatFix&) { record_msg("test/navsatfix"); });
        z_create_subscription<z_JoyFeedback>("test/joyfeedback", [](const z_JoyFeedback&) { record_msg("test/joyfeedback"); });
        z_create_subscription<z_JointState>("test/jointstate", [](const z_JointState&) { record_msg("test/jointstate"); });
        z_create_subscription<z_Vector3>("test/vector3", [](const z_Vector3&) { record_msg("test/vector3"); });
        z_create_subscription<z_Quaternion>("test/quaternion", [](const z_Quaternion&) { record_msg("test/quaternion"); });
        z_create_subscription<z_Pose>("test/pose", [](const z_Pose&) { record_msg("test/pose"); });
        z_create_subscription<z_Twist>("test/twist", [](const z_Twist&) { record_msg("test/twist"); });

        // Custom Topics
        z_create_subscription<z_SetLED>("test_custom/set_led", [](const z_SetLED&) { record_msg("test_custom/set_led"); });
        z_create_subscription<z_MPU6050Data>("test_custom/mpu6050", [](const z_MPU6050Data&) { record_msg("test_custom/mpu6050"); });
        z_create_subscription<z_RobotState>("test_custom/robot_state", [](const z_RobotState&) { record_msg("test_custom/robot_state"); });
        z_create_subscription<z_MotorStatus>("test_custom/motor_status", [](const z_MotorStatus&) { record_msg("test_custom/motor_status"); });
        z_create_subscription<z_SensorTelemetry>("test_custom/telemetry", [](const z_SensorTelemetry&) { record_msg("test_custom/telemetry"); });
        z_create_subscription<z_RobotDiagnostic>("test_custom/robot_diag", [](const z_RobotDiagnostic&) { record_msg("test_custom/robot_diag"); });
        z_create_subscription<z_ComplexRobotState>("test_custom/complex_state", [](const z_ComplexRobotState&) { record_msg("test_custom/complex_state"); });

        // Custom Service Clients
        cli_set_led_clr_ = z_create_client<z_SetLEDColor>("srv_custom/set_led_color");
        cli_set_clr_     = z_create_client<z_SetColor>("srv_custom/set_color");
        cli_cfg_robot_   = z_create_client<z_ConfigureRobot>("srv_custom/configure_robot");
        cli_full_ctrl_   = z_create_client<z_FullSystemControl>("srv_custom/full_system_control");
    }

    bool call_services() {
        bool ok = true;
        // 1. SetLEDColor
        z_SetLEDColor::Request req1; req1.led_data.r = 255; req1.led_data.brightness = 100;
        z_SetLEDColor::Response res1;
        if (!cli_set_led_clr_->call(req1, res1, 1500) || !res1.success) ok = false;

        // 2. SetColor
        z_SetColor::Request req2; req2.r = 255; req2.g = 128; req2.b = 0;
        z_SetColor::Response res2;
        if (!cli_set_clr_->call(req2, res2, 1500) || !res2.success) ok = false;

        // 3. ConfigureRobot
        z_ConfigureRobot::Request req3; req3.target_status.motor_id = 1; req3.target_status.speed = 100.0f; req3.target_status.is_active = true; req3.mode = 1;
        z_ConfigureRobot::Response res3;
        if (!cli_cfg_robot_->call(req3, res3, 1500) || !res3.success) ok = false;

        // 4. FullSystemControl
        z_FullSystemControl::Request req4; req4.target_state.robot_mode = "STRESS_CPP"; req4.command_code = 9;
        z_FullSystemControl::Response res4;
        if (!cli_full_ctrl_->call(req4, res4, 1500) || !res4.ack) ok = false;

        return ok;
    }

private:
    ZenohClient<z_SetLEDColor>* cli_set_led_clr_;
    ZenohClient<z_SetColor>* cli_set_clr_;
    ZenohClient<z_ConfigureRobot>* cli_cfg_robot_;
    ZenohClient<z_FullSystemControl>* cli_full_ctrl_;
};

int main(int argc, char** argv) {
    std::cout << "\n==============================================================\n";
    std::cout << "  CONTINUOUS 2-MINUTE STRESS TEST & VALIDATOR (C++)\n";
    std::cout << "  Testing continuous data stream & frequent service calls\n";
    std::cout << "==============================================================\n\n";

    ZenohConfig cfg;
    cfg.host = (argc > 1) ? argv[1] : "10.42.0.50";
    cfg.port = 7447;

    if (!ZenohNode::init(cfg)) {
        std::cerr << "Zenoh init failed!\n";
        return 1;
    }

    ContinuousStressNode node;

    std::thread spin_thread([&]() {
        node.z_spin();
    });

    auto start_time = std::chrono::steady_clock::now();
    uint32_t total_srv_calls = 0;
    uint32_t successful_srv_calls = 0;
    auto last_srv_time = std::chrono::steady_clock::now();

    while (true) {
        std::this_thread::sleep_for(std::chrono::milliseconds(500));
        auto now = std::chrono::steady_clock::now();
        int elapsed = std::chrono::duration_cast<std::chrono::seconds>(now - start_time).count();

        if (elapsed >= 120) break;

        int srv_elapsed = std::chrono::duration_cast<std::chrono::seconds>(now - last_srv_time).count();
        if (srv_elapsed >= 2) {
            last_srv_time = now;
            total_srv_calls += 4;
            if (node.call_services()) {
                successful_srv_calls += 4;
            }
        }

        if (elapsed % 10 == 0 && elapsed > 0) {
            std::lock_guard<std::mutex> lock(mtx);
            std::cout << "  [" << elapsed << "s/120s] Active Topics: " << topic_counts.size()
                      << "/34 | Total Msgs: " << total_msg_counter.load()
                      << " | Srv Calls OK: " << successful_srv_calls << "/" << total_srv_calls << "\n";
        }
    }

    std::lock_guard<std::mutex> lock(mtx);
    std::cout << "\n==============================================================\n";
    std::cout << "  CONTINUOUS STRESS TEST FINISHED (120s elapsed)\n";
    std::cout << "  Active Topics  : " << topic_counts.size() << "/34\n";
    std::cout << "  Total Messages : " << total_msg_counter.load() << "\n";
    std::cout << "  Service Calls  : " << successful_srv_calls << "/" << total_srv_calls
              << " PASSED (" << (double)successful_srv_calls/total_srv_calls*100.0 << "%)\n";
    if (topic_counts.size() == 34 && successful_srv_calls == total_srv_calls) {
        std::cout << "  \033[32m\033[1m✓ 100% PERFECT CONTINUOUS STABILITY VERIFIED! ZERO LOSS!\033[0m\n";
    } else {
        std::cout << "  \033[31m\033[1m⚠  STRESS TEST COMPLETED WITH LOSS/ERRORS\033[0m\n";
    }
    std::cout << "==============================================================\n\n";

    spin_thread.detach();
    return 0;
}
